### PR TITLE
Add raw public key provisioning.

### PR DESCRIPTION
**Pull request by xiangfeima**
Fix: #1328
This pull request implements DER encoded raw public key parsing and identity authentication (one-way or mutual) when raw public key is provisioned instead of using chertificate chain.

Signed-off-by: Xiangfei Ma <xiangfei.ma@samsung.com>

### CODE REVIEW

File: include/internal/libspdm_common_lib.h
Comment (steven-bellock) [inline include/internal/libspdm_common_lib.h:656]: `public_key_hash_size` isn't needed since this info is in the `spdm_context`.

**Review by steven-bellock — CHANGES_REQUESTED** [review]

File: include/internal/libspdm_common_lib.h
Comment (xiangfeima) [reply include/internal/libspdm_common_lib.h:580]: The name was designed to be aligned with the function we already have for certificate chain hash generation/verification.
`bool libspdm_generate_cert_chain_hash()` <-> `bool libspdm_generate_public_key_hash()`
`bool libspdm_verify_certificate_chain_hash()` <-> `bool libspdm_verify_public_key_hash()`
Since the public key hash generation/verification process has the same routine as certificate chain does (one only **generates local hash** and **verifies peer hash**), so I tend to keep the function name as the way it is right now.

File: include/hal/library/cryptlib/cryptlib_ec.h
Comment (xiangfeima) [reply include/hal/library/cryptlib/cryptlib_ec.h:30]: All "Pointer to new-generated" phrases will be chaged to "Pointer to the newly generated" as suggested by next patch set.

File: include/internal/libspdm_common_lib.h
Comment (xiangfeima) [reply include/internal/libspdm_common_lib.h:656]: There is a comparison between `public_key_hash_size` and the `hash_size` from `spdm_context` in the hash verification process (just as same as certificate chain has verification process does), so the `public_key_hash_size` is necessary.

File: library/spdm_common_lib/libspdm_com_crypto_service.c
Comment (jyao1) [inline library/spdm_common_lib/libspdm_com_crypto_service.c:31]: why you remove this ASSERT?

File: library/spdm_common_lib/libspdm_com_crypto_service.c
Comment (jyao1) [inline library/spdm_common_lib/libspdm_com_crypto_service.c:74]: The logic seems return data if LIBSPDM_RECORD_TRANSCRIPT_DATA_SUPPORT.

If LIBSPDM_RECORD_TRANSCRIPT_DATA_SUPPORT is 0, then it return FALSE.

Is that possible to align libspdm_get_peer_cert_chain_data() to libspdm_get_peer_cert_chain_buffer()?

File: library/spdm_common_lib/libspdm_com_crypto_service.c
Comment (xiangfeima) [reply library/spdm_common_lib/libspdm_com_crypto_service.c:31]: It's a miss. Thanks for pointing it out.
I'll fix it by next patch set.

File: library/spdm_common_lib/libspdm_com_crypto_service.c
Comment (xiangfeima) [reply library/spdm_common_lib/libspdm_com_crypto_service.c:74]: > The logic seems return data if LIBSPDM_RECORD_TRANSCRIPT_DATA_SUPPORT.
> If LIBSPDM_RECORD_TRANSCRIPT_DATA_SUPPORT is 0, then it return FALSE.

Yes indeed.
Because we store peer certificate chain data in context only if LIBSPDM_RECORD_TRANSCRIPT_DATA_SUPPORT is 1.
https://github.com/DMTF/libspdm/blob/e9e1d0a37a42dd3b91fdf63e8e33f038e98e0d09/include/internal/libspdm_common_lib.h#L44-L54

> Is that possible to align libspdm_get_peer_cert_chain_data() to libspdm_get_peer_cert_chain_buffer()?

I lost you here. Would you please explain this comment more clearly?

File: library/spdm_common_lib/libspdm_com_crypto_service.c
Comment (jyao1) [inline library/spdm_common_lib/libspdm_com_crypto_service.c:610]: Originally, this API is to check the runtime data v.s. local_context.

But, the local_context.peer_cert_chain_provision is moved. The check is change to runtime data v.s. connection_info.

I dont believe that is the expectation.

I recommend we remove libspdm_verify_peer_digests() API at all, because there is nothing to verify.

File: library/spdm_common_lib/libspdm_com_crypto_service.c
Comment (xiangfeima) [reply library/spdm_common_lib/libspdm_com_crypto_service.c:74]: I see, the `result` seems to be unnecessary. I'll change the way it reture to be aligned with `libspdm_get_peer_cert_chain_buffer()`.

Comment (jyao1) [reply library/spdm_common_lib/libspdm_com_crypto_service.c:74]: I mean, just change the 
```
libspdm_get_peer_cert_chain_data()
{
#if LIBSPDM_RECORD_TRANSCRIPT_DATA_SUPPORT
    ......
#else
    return FALSE;
#endif
}
```

I dont see the need to define result and assign to FALSE, then return result, if LIBSPDM_RECORD_TRANSCRIPT_DATA_SUPPORT==0

File: library/spdm_common_lib/libspdm_com_crypto_service.c
Comment (jyao1) [inline library/spdm_common_lib/libspdm_com_crypto_service.c:944]: It seems you removed "else" logic.
Do you assume buffer_hash_size is always !=0?
If yes, I suggest to add ASSERT().

Comment (xiangfeima) [reply library/spdm_common_lib/libspdm_com_crypto_service.c:944]: > It seems you removed "else" logic.

Yes. The "else" logic was trying to read provisioned peer certificate chain from context.
It is no longer needed since local_context.peer_cert_chain_provision is gone.

> Do you assume buffer_hash_size is always !=0?
> If yes, I suggest to add ASSERT().

Yes, buffer_hash_size must not be 0 here.
Because this function is called in CHALLENGE process,
the certificate chain hash must be there from GET_CERTIFICATE process or imported from context cache.
I'll add the assertion.

File: library/spdm_responder_lib/libspdm_rsp_key_exchange.c
Comment (jyao1) [inline library/spdm_responder_lib/libspdm_rsp_key_exchange.c:256]: Since slot_id maybe 0xFF, do we need similar check above?
E.g. check to ensure spdm_context->local_context.local_public_key_provision

File: library/spdm_responder_lib/libspdm_rsp_challenge_auth.c
Comment (jyao1) [inline library/spdm_responder_lib/libspdm_rsp_challenge_auth.c:162]: Since slot_id maybe 0xFF, do we need similar check above line 89?
E.g. check to ensure spdm_context->local_context.local_public_key_provision

File: os_stub/cryptlib_mbedtls/der/der.c
Comment (jyao1) [inline os_stub/cryptlib_mbedtls/der/der.c:8]: Is that possible to add some crypto unit test for the new APIs in https://github.com/DMTF/libspdm/tree/main/unit_test/test_crypt?

File: unit_test/test_spdm_requester/encap_request.c
Comment (jyao1) [inline unit_test/test_spdm_requester/encap_request.c:76]: why we need change this? Is it related to raw public key provisioning?

File: unit_test/test_spdm_requester/encap_request.c
Comment (jyao1) [inline unit_test/test_spdm_requester/encap_request.c:98]: why we need change this? Is it related to raw public key provisioning?

**Comment by jyao1** [comment]
@xiangfeima , thanks to contribute this patch. I think the overall design is clear.
I have finished 1st round review and provided all feedback for the implementation.


Since the code change is big, I would like to double check if there is enough validation for the feature.
1) enable spdm_emu to test 0xFF (I think you already have that, great)
2) enable spdm_dump (https://github.com/DMTF/spdm-dump) to test 0xFF (I dont see PR there).
3) test new spdm_requester_emu(slot - FF) with new spdm_responder_emu(slot - FF) - I assume you have done.
4) test new spdm_requester_emu(slot - non-FF) with new spdm_responder_emu(slot - non-FF) - I assume you have done.
5) test old spdm_requester_emu(slot - non-FF) with new spdm_responder_emu(slot - non-FF) - this is to ensure change will not break the existing transcript.
6) test new spdm_requester_emu(slot - non-FF) with old spdm_responder_emu(slot - non-FF) - this is to ensure change will not break the existing transcript.

File: library/spdm_common_lib/libspdm_com_crypto_service.c
Comment (xiangfeima) [reply library/spdm_common_lib/libspdm_com_crypto_service.c:610]: In my opinion the libspdm_verify_peer_digests() was supposed to be checking runtime retrieved digests data.
Refer to SPDM specification v1.2 `Figure 8 — Responder authentication: Example certificate retrieval flow`, 
> 2. Compare digests in DIGESTS response
message to cached digests. _**Continue if
no match is found**_.

The word **_Continue_** here in my understanding is to perform `GET_CERTIFICATE` afterwards.
Thus as I see the use cases of libspdm_verify_peer_digests() are still there:
1. First time `GET_DIGESTS`. The libspdm_verify_peer_digests() should return FALSE (since there is nothing in local context to verify with the retrieved digests), indicate people to perform `GET_CERTIFICATE` (Continue if
no match is found), and cache digest of certificate chain per slot in local context connection_info by the `GET_CERTIFICATE`.
2. `GET_DIGESTS` after `LIBSPDM_CONNECTION_STATE_AUTHENTICATED`. The libspdm_verify_peer_digests() should return actual check result (runtime retrieved digests data v.s. cached digest per slot in local context connection_info), to tell if the peer identity is changed (after a `SET_CERTIFICATE`) or tampered, and a `GET_CERTIFICATE` afterwards may be needed base upon the check result.

Currently `GET_CERTIFICATE` is called only if `GET_DIGESTS` is succeeded so I keep the libspdm_verify_peer_digests() to return TRUE at first time it is called (nothing is cached in local context connection_info). It can be filed to another issue and be resolved later.
https://github.com/DMTF/spdm-emu/blob/9202e51b57abba044387746a926832df8174114d/spdm_emu/spdm_requester_emu/spdm_requester_authentication.c#L42-L54

In summary, I think libspdm_verify_peer_digests() is needed.

File: library/spdm_responder_lib/libspdm_rsp_key_exchange.c
Comment (xiangfeima) [reply library/spdm_responder_lib/libspdm_rsp_key_exchange.c:256]: No need to.
The slot_id is clearly checked just at the line L224-L239 above.

File: library/spdm_responder_lib/libspdm_rsp_challenge_auth.c
Comment (xiangfeima) [reply library/spdm_responder_lib/libspdm_rsp_challenge_auth.c:162]: No need to.
The slot_id is clearly checked just at the line L81-L95 above.

File: os_stub/cryptlib_mbedtls/der/der.c
Comment (xiangfeima) [reply os_stub/cryptlib_mbedtls/der/der.c:8]: Sure. I will add TCs accordingly.

File: include/hal/library/cryptlib/cryptlib_rsa.h
Comment (jyao1) [inline include/hal/library/cryptlib/cryptlib_rsa.h:39]: DER is just encoding.

I think we need clarify which standard we are following for the RSA public key format.

File: unit_test/test_spdm_requester/encap_request.c
Comment (xiangfeima) [reply unit_test/test_spdm_requester/encap_request.c:76]: This is changed because both of
1. provisioned peer certificate chain in local context is gone
2. the way that libspdm_verify_peer_digests() is changed

Previously the way of testing
> send all eight certchains digest but only No.7 is right
would be succeeded because we were doing it in the way that to see if there is digest of one slot in retrieved digests data matches the hash of provisioned peer chetificate chain in local context.

Since provisioned peer certificate chain in local context is gone,
and we are caching peer certificate chain data/hash slot by slot in `GET_CERTIFICATE` process,
the way to check peer digests in libspdm_verify_peer_digests() is also changed to check the hash value of every slot according to given slot_mask.

And due to that, `send all eight certchains digest but only No.7 is right` way of testing won't be succeeded any more.
Now pass condition for `GET_DIGESTS` should be: hash value for every slot is correct against the given slot_mask.

File: unit_test/test_spdm_requester/encap_request.c
Comment (xiangfeima) [reply unit_test/test_spdm_requester/encap_request.c:98]: This is changed because both of

provisioned peer certificate chain in local context is gone
the way that libspdm_verify_peer_digests() is changed
Previously the way of testing

send all eight certchains digest but only No.7 is right

would be succeeded because we were doing it in the way that to see if there is digest of one slot in retrieved digests data matches the hash of provisioned peer chetificate chain in local context.

Since provisioned peer certificate chain in local context is gone,
and we are caching peer certificate chain data/hash slot by slot in GET_CERTIFICATE process,
the way to check peer digests in libspdm_verify_peer_digests() is also changed to check the hash value of every slot according to given slot_mask.

And due to that, send all eight certchains digest but only No.7 is right way of testing won't be succeeded any more.
Now pass condition for GET_DIGESTS should be: hash value for every slot is correct against the given slot_mask

File: include/hal/library/cryptlib/cryptlib_rsa.h
Comment (xiangfeima) [reply include/hal/library/cryptlib/cryptlib_rsa.h:39]: The description is added to doc/user_guide.md
> The public key is ASN.1 DER-encoded as [RFC7250](https://www.rfc-editor.org/rfc/rfc7250) describes,
    namely, the `SubjectPublicKeyInfo` structure of a X.509 certificate.

I can add such comment to here as well if you insist.

**Comment by xiangfeima** [comment]
Hi @jyao1 , thanks for your time to go through such many code change.
About the 6 points you listed above for the feature validation, I will update the result here alone with I update next patch set to fix the comments from 1st round review.

File: library/spdm_responder_lib/libspdm_rsp_key_exchange.c
Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_key_exchange.c:256]: See comment in https://github.com/DMTF/libspdm/pull/1556#discussion_r1062308490

Comment (xiangfeima) [reply library/spdm_responder_lib/libspdm_rsp_key_exchange.c:256]: Point is taken, I will fix it by next patch set.

File: library/spdm_responder_lib/libspdm_rsp_challenge_auth.c
Comment (xiangfeima) [reply library/spdm_responder_lib/libspdm_rsp_challenge_auth.c:162]: Point is taken, I will fix it by next patch set.

File: library/spdm_common_lib/libspdm_com_crypto_service.c
Comment (jyao1) [reply library/spdm_common_lib/libspdm_com_crypto_service.c:610]: > First time GET_DIGESTS. The libspdm_verify_peer_digests() should return FALSE

This is not the purpose, when we designed libspdm_verify_peer_digests().
The expectation of verify_xxx to check if the data from the peer meets the basic requirement before return the consumer.
As such, we only verify if it matches the provisioned data. If no data is provisioned, this API should just return TRUE, because there is nothing to verify.

> GET_DIGESTS after LIBSPDM_CONNECTION_STATE_AUTHENTICATED. The libspdm_verify_peer_digests() should return actual check result

This is also not the purpose of verify_xxx API. Thinking of runtime update use case, the change of digest is a valid condition. That helps the consumer to know that the cached digest is no longer valid. The requester should send get_certificate again to get the latest result. 



To me, what you describe above is something like is_cached_data_valid() for runtime data, instead of verify_xxx() according to the provisioned data.
My suggestion is: Don't mix them in raw public key provision patch. Just remove libspdm_verify_peer_digests() in this patch. If you really want to add is_cached_data_valid(), let's use another issue and PR.

Comment (xiangfeima) [reply library/spdm_common_lib/libspdm_com_crypto_service.c:610]: This explanation helps a lot, thank you.
Now I fully understand the design of `libspdm_verify_peer_digests()` is only to verify if the retrieved digest matches provisioned data before return it to consumer.
I will remove `libspdm_verify_peer_digests()` by next patch set as you suggested since it is no longer needed.

Also, there's not much need to do such `is_cached_data_valid()` thing since it would be consumer's call to send get_certificate (or not) afterwards.

File: unit_test/test_spdm_requester/encap_request.c
Comment (xiangfeima) [reply unit_test/test_spdm_requester/encap_request.c:76]: I will remain the way to generate digest in test routine as the way it is right now (only send correct digest value of one slot for most cases), because the previous way of digest generation
> send all eight certchains digest but only No.7 is right

since `libspdm_verify_peer_digests()` will be removed at all.
Refer to https://github.com/DMTF/libspdm/pull/1556#discussion_r1063003386

File: unit_test/test_spdm_requester/encap_request.c
Comment (xiangfeima) [reply unit_test/test_spdm_requester/encap_request.c:98]: I will remain the way to generate digest in test routine as the way it is right now (only send correct digest value of one slot for most cases), because the previous way of digest generation
> send all eight certchains digest but only No.7 is right

is no longer needed, since `libspdm_verify_peer_digests()` will be removed at all.
Refer to https://github.com/DMTF/libspdm/pull/1556#discussion_r1063003386

**Comment by xiangfeima** [comment]
@jyao1 , all comments from your 1st round review are fixed by https://github.com/DMTF/libspdm/commit/44dde488c819719ab6571450403f2cbd5d9539d7.

The status of items for feature validation are updated as below:
1. enable spdm_emu to test 0xFF  **-> https://github.com/DMTF/spdm-emu/pull/204**
2. enable spdm_dump (https://github.com/DMTF/spdm-dump) to test 0xFF **-> In progress**
3. test new spdm_requester_emu(slot - FF) with new spdm_responder_emu(slot - FF) **-> Tested and confirmed working well**
4. test new spdm_requester_emu(slot - non-FF) with new spdm_responder_emu(slot - non-FF) **-> Tested and confirmed working well**
5. test old spdm_requester_emu(slot - non-FF) with new spdm_responder_emu(slot - non-FF) - this is to ensure change will not break the existing transcript. **-> Tested and confirmed working well**
6. test new spdm_requester_emu(slot - non-FF) with old spdm_responder_emu(slot - non-FF) - this is to ensure change will not break the existing transcript. **-> Tested and confirmed working well**

**Review by jyao1 — APPROVED** [review]

**Comment by jyao1** [comment]
Thanks @xiangfeima, I approved it.

Once you finish 2) the spdm_dump, we can merge it.

**Comment by jyao1** [comment]
@xiangfeima, I will merge this now, since we have more pending P-R, which may cause conflict.

Please continue working on spdm_dump and submit patch there. Thank you!

**Comment by xiangfeima** [comment]
@jyao1 thank you. I'll update the patch to spdm-dump soon.
(The patch itself is ready, the last thing I'm working on is to finish some internal compliance process.)

**Comment by xiangfeima** [comment]
@jyao1 The patch for spdm-dump is update to https://github.com/DMTF/spdm-dump/pull/75

**Comment by jyao1** [comment]
All merged

**Comment by xiangfeima** [comment]
Thank you 👍

**Pull request merged**
